cipher-bouncy/src/main/scala/tsec/cipher/symmetric/bouncy/ChaCha20Poly1305.scala

```diff
@@ -0,0 +1,270 @@
+          key: BouncySecretKey[ChaCha20Poly1305],
+          iv: Iv[ChaCha20Poly1305],
+          aad: AAD
+      ): F[CipherText[ChaCha20Poly1305]] =
+        F.delay(impl.unsafeEncryptAAD(plainText, key, iv, aad))
+
+      def encryptWithAADDetached(
+          plainText: PlainText,
+          key: BouncySecretKey[ChaCha20Poly1305],
+          iv: Iv[ChaCha20Poly1305],
+          aad: AAD
+      ): F[(CipherText[ChaCha20Poly1305], AuthTag[ChaCha20Poly1305])] =
+        F.delay(impl.unsafeEncryptDetachedAAD(plainText, key, iv, aad))
+
+      def decryptWithAAD(
+          cipherText: CipherText[ChaCha20Poly1305],
+          key: BouncySecretKey[ChaCha20Poly1305],
+          aad: AAD
+      ): F[PlainText] =
+        F.delay(impl.unsafeDecryptAAD(cipherText, key, aad))
+
+      def decryptWithAADDetached(
+          cipherText: CipherText[ChaCha20Poly1305],
+          key: BouncySecretKey[ChaCha20Poly1305],
+          aad: AAD,
+          authTag: AuthTag[ChaCha20Poly1305]
+      ): F[PlainText] =
+        F.delay(impl.unsafeDecryptDetachedAAD(cipherText, authTag, key, aad))
+
+      def encryptDetached(
+          plainText: PlainText,
+          key: BouncySecretKey[ChaCha20Poly1305],
+          iv: Iv[ChaCha20Poly1305]
+      ): F[(CipherText[ChaCha20Poly1305], AuthTag[ChaCha20Poly1305])] =
+        F.delay(impl.unsafeEncryptDetached(plainText, key, iv))
+
+      def decryptDetached(
+          cipherText: CipherText[ChaCha20Poly1305],
+          key: BouncySecretKey[ChaCha20Poly1305],
+          authTag: AuthTag[ChaCha20Poly1305]
+      ): F[PlainText] =
+        F.delay(impl.unsafeDecryptDetached(cipherText, authTag, key))
+
+      def encrypt(
+          plainText: PlainText,
+          key: BouncySecretKey[ChaCha20Poly1305],
+          iv: Iv[ChaCha20Poly1305]
+      ): F[CipherText[ChaCha20Poly1305]] =
+        F.delay(impl.unsafeEncrypt(plainText, key, iv))
+
+      def decrypt(cipherText: CipherText[ChaCha20Poly1305], key: BouncySecretKey[ChaCha20Poly1305]): F[PlainText] =
+        F.delay(impl.unsafeDecrypt(cipherText, key))
+    }
+
+  def defaultIvGen[F[_]](implicit F: Sync[F]): IvGen[F, ChaCha20Poly1305] =
+    new IvGen[F, ChaCha20Poly1305] with ManagedRandom {
+
+      def genIv: F[Iv[ChaCha20Poly1305]] =
+        F.delay(genIvUnsafe)
+
+      def genIvUnsafe: Iv[ChaCha20Poly1305] = {
+        val nonce = new Array[Byte](NonceLenBytes)
+        nextBytes(nonce)
+        Iv[ChaCha20Poly1305](nonce)
+      }
+    }
+
+  object impl {
+    def unsafeEncrypt(
+        plainText: PlainText,
+        k: BouncySecretKey[ChaCha20Poly1305],
+        iv: Iv[ChaCha20Poly1305]
+    ): CipherText[ChaCha20Poly1305] =
+      unsafeEncryptAAD(plainText, k, iv, AAD(Array.empty[Byte]))
+
+    def unsafeEncryptAAD(
+        plainText: PlainText,
+        k: BouncySecretKey[ChaCha20Poly1305],
+        iv: Iv[ChaCha20Poly1305],
+        aad: AAD
+    ): CipherText[ChaCha20Poly1305] = {
+      val chacha20   = new ChaChaEngine(20)
```

Comment:
Can we add a link to a document describing all of these numbers ?

Reply:
I can yeah. Cryptographic code is pretty fucking full of magic numbers.